File: _includes/sort-benchmark/algorithms/pairing_heap.rs
struct PairingNode {
    key: usize,
    child: Option<Box<PairingNode>>,
    sibling: Option<Box<PairingNode>>,
}

fn meld(
    a: Option<Box<PairingNode>>,
    b: Option<Box<PairingNode>>,
) -> Option<Box<PairingNode>> {
    match (a, b) {
        (None, x) | (x, None) => x,
        (Some(mut x), Some(mut y)) => {
            if x.key <= y.key {
                y.sibling = x.child.take();
                x.child = Some(y);
                Some(x)
            } else {
                x.sibling = y.child.take();
                y.child = Some(x);
                Some(y)
            }
        }
    }
}

fn two_pass_meld(mut first: Option<Box<PairingNode>>) -> Option<Box<PairingNode>> {
    let mut pairs = Vec::new();
    while let Some(mut a) = first.take() {
        first = a.sibling.take();
        if let Some(mut b) = first.take() {
            first = b.sibling.take();
            pairs.push(meld(Some(a), Some(b)));
        } else {
            pairs.push(Some(a));
        }
    }

    let mut result = None;
    for pair in pairs.into_iter().rev() {
        result = meld(pair, result);
    }
    result
}

fn insert_key(heap: Option<Box<PairingNode>>, key: usize) -> Option<Box<PairingNode>> {
    let node = Box::new(PairingNode {
        key,
        child: None,
        sibling: None,
    });
    meld(heap, Some(node))
}

fn extract_min(heap: Option<Box<PairingNode>>) -> (Option<usize>, Option<Box<PairingNode>>) {
    let Some(mut root) = heap else {
        return (None, None);
    };
    let key = root.key;
    let children = root.child.take();
    (Some(key), two_pass_meld(children))
}

fn pairing_heap_sort(a: &mut [usize]) {
    let mut heap = None;
    for &key in a.iter() {
        heap = insert_key(heap, key);
    }
    for slot in a.iter_mut() {
        let (key, next) = extract_min(heap);
        heap = next;
        *slot = key.expect("pairing heap exhausted early");
    }
}

File: _includes/sort-benchmark/algorithms/pairing_heap.swift
fileprivate final class PairingNode {
    var key: Int
    var child: PairingNode?
    var sibling: PairingNode?

    init(key: Int, child: PairingNode? = nil, sibling: PairingNode? = nil) {
        self.key = key
        self.child = child
        self.sibling = sibling
    }
}

fileprivate func pairing_meld(_ a: PairingNode?, _ b: PairingNode?) -> PairingNode? {
    guard let a else { return b }
    guard let b else { return a }
    if a.key <= b.key {
        b.sibling = a.child
        a.child = b
        return a
    } else {
        a.sibling = b.child
        b.child = a
        return b
    }
}

fileprivate func pairing_two_pass_meld(_ first: PairingNode?) -> PairingNode? {
    var first = first
    var pairs = [PairingNode?]()
    while let a = first {
        first = a.sibling
        a.sibling = nil
        if let b = first {
            first = b.sibling
            b.sibling = nil
            pairs.append(pairing_meld(a, b))
        } else {
            pairs.append(a)
        }
    }

    var result: PairingNode? = nil
    for pair in pairs.reversed() {
        result = pairing_meld(pair, result)
    }
    return result
}

fileprivate func pairing_insert_key(_ heap: PairingNode?, _ key: Int) -> PairingNode? {
    let node = PairingNode(key: key)
    return pairing_meld(heap, node)
}

fileprivate func pairing_extract_min(_ heap: PairingNode?) -> (Int?, PairingNode?) {
    guard let root = heap else {
        return (nil, nil)
    }
    let key = root.key
    let children = root.child
    root.child = nil
    return (key, pairing_two_pass_meld(children))
}

func pairing_heap_sort(_ a: inout [Int]) {
    a.withUnsafeMutableBufferPointer { pairing_heap_sort($0) }
}

func pairing_heap_sort(_ a: UnsafeMutableBufferPointer<Int>) {
    var heap: PairingNode? = nil
    for i in 0..<a.count {
        heap = pairing_insert_key(heap, a[i])
    }
    for i in 0..<a.count {
        let (key, next) = pairing_extract_min(heap)
        heap = next
        guard let key else {
            fatalError("pairing heap exhausted early")
        }
        a[i] = key
    }
}
